include/bluetooth/services/ots.h

```diff
@@ -0,0 +1,177 @@
+			/** Bit 4 Appending data to this object that increases
+			 *	  its Allocated Size is permitted
+			 */
+			bool append_permitted	: 1;
+			/** Bit 5 Truncation of this object is permitted */
+			bool truncate_permitted	: 1;
+			/** Bit 6 Patching this object by overwriting some
+			 *	  of the object's existing contents is permitted
+			 */
+			bool patch_permitted	: 1;
+			/** Bit 7 This object is a marked object */
+			bool is_marked		: 1;
+		} decoded;
+
+		/* Properties as raw value */
+		u32_t raw;
+	};
+};
+
+/**@brief Descriptor for OTS object intialization. */
+struct bt_gatt_ots_obj_init {
+	/* Object Name */
+	char                        *name;
+
+	/* Object Type */
+	struct bt_gatt_ots_obj_type  type;
+
+	/* Object Size (allocated and current) */
+	u32_t                        size;
+
+	/* Object Properties */
+	struct bt_gatt_ots_obj_props props;
+};
+
+/** @brief OTS server callback structure. */
+struct bt_gatt_ots_cb {
+	/** Object created callback
+	 *
+	 *  This callback is mandatory. It is called whenever a new object is
+	 *  created. object data should be set within the context of this
+	 *  callback.
+	 *
+	 *  @param conn The connection that is requesting object creation or
+	 *              NULL if object is created by bt_gatt_ots_object_add()
+	 *  @param id   Object ID
+	 *  @param data Pointer to receive the buffer to object data
+	 *  @param size Allocated Size of the object
+	 *
+	 *  @return 0 in case of success or negative value in case of error.
+	 *  Possible return values:
+	 *  -ENOMEM if no available space for new object.
+	 */
+	int (*obj_created)(struct bt_conn *conn, u64_t id, u8_t **data,
+			   u32_t size);
+
+	/** Object deleted callback
+	 *
+	 *  This callback is called whenever an object is deleted.
+	 *
+	 *  @param conn The connection that deleted the object or NULL if
+	 *              this request came from the server
+	 *  @param id   Object ID
+	 *  @param data Object data
+	 */
+	void (*obj_deleted)(struct bt_conn *conn, u64_t id, u8_t *data);
+
+	/** Object selected callback
+	 *
+	 *  This callback is called on successful object selection.
+	 *
+	 *  @param conn The connection that selected new object
+	 *  @param id   Object ID
+	 */
+	void (*obj_selected)(struct bt_conn *conn, u64_t id);
+
+	/** Object read callback
+	 *
+	 *  This callback is called on successful object read.
+	 *
+	 *  @param conn The connection that read object
+	 *  @param id   Object ID
+	 */
+	void (*obj_read)(struct bt_conn *conn, u64_t id);
+};
+
+/** @brief Add an object to the OTS Server.
+ *
+ *  This function adds an object to the OTS database. When the
+ *  object is being added, a callback obj_created() is called
+ *  to notify the user about new object ID and to receive pointer
+ *  to object data.
+ *
+ *  @param obj_init Meta data of the object
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_gatt_ots_object_add(struct bt_gatt_ots_obj_init *obj_init);
```

Comment:
I've also been thinking about this, and what the best course of action is. 

On one side we have a simple pointer that is assigned when an object is created (either by the server or the client). This approach is simple, and doesn't require the application to deal with any reads or writes. 

On the other side we have the callback approach, where it is, as @leifAlex mentions, possible for the application to handle the data in other ways. 

So should we have a data pointer, or a callback for this? Or possibly both?

Reply:
Well, I think it's best to pick one way of handling the read operation.

If we want the application to control the transfer process, then I will implement the chunk-based approach (with callback). I think it could be also used to cover the functionality mentioned by @Thalley in another comment: to inform the service user about the start and end of the read operation.

In this case, I will get rid of ``data`` pointer in the ``obj_created`` callback.

Is it a good way forward for you? Do you have any comments?

Reply:
In the IOP implementation I had done, I've had both way being possible, where I would check if there is a 'read_cb' provided then call it, otherwise use the data from the pointer given at init time (and if that one is null, have an error). Not really sure it's needed to keep both options, makes it potentially confusing for end user...

Reply:
I think it would be best to have a single callback. Either a single callback on the OTS instance with object ID, or, as Leif implemented, a callback for each object.